### ISSUE TITLE
Why doing one update per rollout_thread?

### ISSUE THREAD

**Issue by julienroyd** [shariqiqbal2810/maddpg-pytorch#7]
Thanks for the great code!

I've got more of a question than an issue: in main.py you do one update per rollout thread
```
for u_i in range(config.n_rollout_threads):
        for a_i in range(maddpg.nagents):
                sample = replay_buffer.sample(config.batch_size, to_gpu=USE_CUDA)
                maddpg.update(sample, a_i, logger=logger)
        maddpg.update_all_targets()
```
And then in the update method for MADDPG agents, you average the gradients:
```
if parallel:
        average_gradients(curr_agent.critic)
```
Why doing that rather than just a single update? 

You use synchronous rollout threads that wait for each other at each timestep, and put all of their collected transitions in a single buffer. You then train using the master-thread only on this single buffer. Since your condition to do an update is to do one every `config.steps_per_update`, the way I see it, it would be less wasteful to just do a single update when this condition is True, and the fact that you have multiple rollout threads would simply help collect the required transitions faster.

Am I missing something?

**Comment by shariqiqbal2810** [comment]
Hi,

So a couple things: 1) the parallel flag is False by default and I'm not setting it to True, so the gradients aren't being averaged. I added that when I was experimenting with parallel training but it wasn't working very well if I remember correctly. 2) I am not sure why I am updating n_rollout_threads times. I think I set it up that way when I was iterating t by 1 each step, but since I am iterating t by n_rollout_threads, it no longer makes sense. It's not necessarily wasteful though since a new sample is being used each iteration so the model it just learning more per update cycle. It does end up doing more updates than is specified by the parameters though, which could be misleading. I'm just going to leave the code as is though since it was working.

**Comment by julienroyd** [comment]
Oh yeah I missed that `parallel` is always False, I guess that's what confused me since we were talking about synchronous learning. I agree it's not worth changing the code and have to re-test everything (for the n_rollout updates), I was just wondering if I was missing something important on the theoretical side. 

Closing the issue.

Thanks again!
